modules/images/webp-uploads/load.php

```diff
@@ -340,3 +340,112 @@ function webp_uploads_wp_get_missing_image_subsizes( $missing_sizes, $image_meta
 }
 
 add_filter( 'wp_get_missing_image_subsizes', 'webp_uploads_wp_get_missing_image_subsizes', 10, 3 );
+
+/**
+ * Filters on `the_content` to update the references for supported mime of images into the
+ * `webp_uploads_preferred_mime_type()` for the most part `image/webp` if the current
+ * attachment contains the targeted mime type.
+ *
+ * @since n.e.x.t
+ *
+ * @see the_content
+ *
+ * @param string $content The content of the current post.
+ * @return string The content with the updated references to the images.
+ */
+function webp_uploads_update_image_references( $content ) {
+	// This content does not have any tag on it, move forward.
+	if ( ! preg_match_all( '/<(img)\s[^>]+>/', $content, $img_tags, PREG_SET_ORDER ) ) {
+		return $content;
+	}
+
+	$images = array();
+	foreach ( $img_tags as list( $img ) ) {
+		// Find the ID of each image by the class.
+		if ( ! preg_match( '/wp-image-([\d]+)/i', $img, $class_name ) ) {
+			continue;
+		}
+
+		if ( empty( $class_name ) ) {
+			continue;
+		}
+
+		// Make sure we use the last item on the list of matches.
+		$attachment_id = (int) end( $class_name );
+
+		if ( ! $attachment_id ) {
+			continue;
+		}
+
+		// Create an array if attachment_id has not been added to the images array.
+		if ( empty( $images[ $attachment_id ] ) ) {
+			$images[ $attachment_id ] = array();
+		}
+
+		$target_extensions = array(
+			'jpeg',
+			'jpg',
+		);
+
+		// Creates a regular extension to find all the URLS with the provided extension for img tag.
+		preg_match_all( '/[^\s"]+\.(?:' . implode( '|', $target_extensions ) . ')/i', $img, $matches );
+
+		$urls = empty( $matches ) ? array() : reset( $matches );
+
+		foreach ( $urls as $url ) {
+			$images[ $attachment_id ][ $url ] = true;
+		}
+	}
+
+	// TODO: Add a filterable option to change the selected mime type. See https://github.com/WordPress/performance/issues/187.
+	$target_mime = 'image/webp';
+	$replacement = array();
+	foreach ( $images as $attachment_id => $urls ) {
+		$metadata = wp_get_attachment_metadata( $attachment_id );
+
+		if ( empty( $metadata['file'] ) ) {
+			continue;
+		}
+
+		$basename = wp_basename( $metadata['file'] );
+
+		foreach ( $urls as $url => $exists ) {
+
+			if ( isset( $metadata['file'] ) && strpos( $url, $basename ) !== false ) {
+				// TODO: we don't have a replacement for full image yet, issue. See: https://github.com/WordPress/performance/issues/174.
+				continue;
+			}
+
+			if ( empty( $metadata['sizes'] ) ) {
+				continue;
+			}
+
+			$src_filename = wp_basename( $url );
+			$extension    = wp_check_filetype( $src_filename );
+
+			// Extension was not set properly no action possible or extension is already in the expected mime.
+			if ( empty( $extension['type'] ) || $extension['type'] === $target_mime ) {
+				continue;
+			}
+
+			// Find the appropriate size for the provided URL.
+			foreach ( $metadata['sizes'] as $name => $size_data ) {
```

Comment:
Actually, this approach would take `O(3N*M)` the current approach is `N*M` core approach is very close the main difference is the fact that core uses functions to hide the complexity see `wp_image_src_get_dimensions` is called for every image which internally triggers a loop for every single size. 

The fact that we iterate on every single size 2 more times 1 to loop metadata and another during the `array_filter` makes the current approach complete in fewer steps. We can do this because we are using a hash map where we know each iteration is unique and we are do not repeat the same step for the same attachment id.

I would split the code into 2 functions as you described above to see if that would make things a bit easier to visualize. But let me know otherwise.

Reply:
> Actually, this approach would take `O(3N*M)`

What is N and M in this, could you clarify? I was specifically referring to the two `foreach` loops in your code here, not in comparison to how it's today in WP core or anything. Good point on considering the `array_filter`, but still from my understanding that would still mean less complexity.

Let's keep this open and revisit after you've split into two functions, that we should definitely do to align the high-level implementation with core so that it can be integrated later.

Reply:
> What is N and M in this, could you clarify? I was specifically referring to the two foreach loops in your code here, not in comparison to how it's today in WP core or anything. Good point on considering the array_filter, but still from my understanding that would still mean less complexity.

M - Being the number of images processed
M - The number of sizes that we need to look for each image in might be the case that both are the same size but technically those two for each are different in terms of the number of sizes.

> Let's keep this open and revisit after you've split into two functions, that we should definitely do to align the high-level implementation with core so that it can be integrated later.

Sure just updated the code based on your feedback let me know in case you have additional concers.